File: src/main/java/com/ggardet/doctor/DoctorServiceInterface.java
package com.ggardet.doctor;

import com.ggardet.doctor.internal.entity.Doctor;
import com.ggardet.doctor.internal.entity.Qualification;
import com.ggardet.doctor.internal.entity.Specialty;

import java.util.List;
import java.util.UUID;

public interface DoctorServiceInterface {
    Doctor createDoctor(final Doctor doctor);

    Doctor findDoctorById(final UUID doctorId);

    List<Doctor> findAllByHospital(final UUID hospitalId);

    List<Doctor> findAllBySpecialty(final String specialty);

    void deleteDoctor(final UUID doctorId);

    void unlinkDoctorFromHospital(final Doctor doctor);

    Specialty createSpecialty(final Specialty specialty);

    Specialty findSpecialtyById(final UUID specialtyId);

    List<Specialty> findAllSpecialties();

    Qualification addQualificationToDoctor(final UUID doctorId, final Qualification qualification);

    List<Qualification> findQualificationsByDoctor(final UUID doctorId);

    void deleteQualification(final UUID qualificationId);
}

File: src/main/kotlin/com/ggardet/doctor/DoctorServiceInterface.kt
package com.ggardet.doctor

import com.ggardet.doctor.internal.entity.Doctor
import com.ggardet.doctor.internal.entity.Qualification
import com.ggardet.doctor.internal.entity.Specialty
import java.util.*

interface DoctorServiceInterface {
    fun createDoctor(doctor: Doctor): Doctor
    
    fun findDoctorById(doctorId: UUID): Doctor
    
    fun findAllByHospital(hospitalId: UUID): List<Doctor>
    
    fun findAllBySpecialty(specialty: String): List<Doctor>
    
    fun deleteDoctor(doctorId: UUID)
    
    fun unlinkDoctorFromHospital(doctor: Doctor)
    
    fun createSpecialty(specialty: Specialty): Specialty
    
    fun findSpecialtyById(specialtyId: UUID): Specialty
    
    fun findAllSpecialties(): List<Specialty>
    
    fun addQualificationToDoctor(doctorId: UUID, qualification: Qualification): Qualification
    
    fun findQualificationsByDoctor(doctorId: UUID): List<Qualification>
    
    fun deleteQualification(qualificationId: UUID)
}